### PR TITLE
Fixes #18104 - Ad hoc Ansible role execution

### PR DESCRIPTION
**Pull request by Gerrit91**
Especially when using Foreman as a backend service, there is yet no
possibility to assign roles to hosts over the API. In addition to that,
some scenarios (like maintenance tasks) require only one-time-execution
of a role or rare role execution from time to time. The need to change
role assignments for each host to do these kind of tasks is impractical.
Therefore, an API where there is no need of specific role assignments
and where an imported role can be executed directly against a certain
host helps to broaden the scopes of application of the Foreman Ansible
plugin. Hopefully, the community is also interested in such
functionality.

### CODE REVIEW

**Comment by Gerrit91** [comment]
Tried to resolve some rubocop issues. :)

**Comment by Gerrit91** [comment]
@dLobatog I will rebase this to the master and will try to keep all your good work from #73. Let's see how long this takes to resolve the issues.

**Comment by Gerrit91** [comment]
I would like to once more advertise this feature. The fact that roles cannot be assigned over the API makes the Ansible Plugin unsuitable for Foreman being used as a pure backend service. Unassigning and reassigning roles to hosts for specific executing a different set of configuration tasks is also quite exhausting.

**Comment by chilicat** [comment]
@dLobatog any opinion to this feature?

File: app/helpers/foreman_ansible/hosts_helper_extensions.rb [hunk 1]
```diff
@@ -13,22 +13,34 @@ def ansible_roles_present?(host)
...
-        icon_text('play', ' ' + _('Ansible roles'), :kind => 'fa'),
-        play_roles_host_path(:id => host.id),
-        :id => :ansible_roles_button,
-        :class => 'btn btn-default',
-        :'data-no-turbolink' => true
-      )
+    def ad_hoc_role_button
+      [link_to_function(
+        icon_text('play', ' ' + _('Ad hoc Ansible role'), :kind => 'fa'),
+        'show_ad_hoc_role_modal()',
```
Comment (xprazak2) [inline app/helpers/foreman_ansible/hosts_helper_extensions.rb]: I am getting `Uncaught ReferenceError: show_ad_hoc_role_modal is not defined` when clicking the button

File: app/controllers/foreman_ansible/api/v2/hostgroups_controller_extensions.rb [hunk 2]
```diff
@@ -45,10 +45,36 @@ def multiple_play_roles
...
+          api :POST, '/hostgroups/:id/play_ad_hoc_role',
+              N_('Plays an Ansible role ad hoc')
+          param :id, :identifier, :required => true
+          param :the_role_id, String, :required => true
+
+          def play_ad_hoc_role
+            # FIXME: When using just role_id, find resource will throw an
+            # exception: "undefined method `name' for nil:NilClass"
+            find_resource
+            find_ansible_role(params.require(:the_role_id))
```
Comment (xprazak2) [inline app/controllers/foreman_ansible/api/v2/hostgroups_controller_extensions.rb]: You can use [parameter_filter](https://github.com/theforeman/foreman_discovery/blob/develop/lib/foreman_discovery/engine.rb#L159-L162) to allow parameters you need

File: app/helpers/foreman_ansible/hosts_helper_extensions.rb [hunk 3]
```diff
@@ -13,22 +13,34 @@ def ansible_roles_present?(host)
...
+      else
+        []
       end
+    end
+
+    def host_title_actions_with_run_ansible_roles(*args)
+      buttons = ad_hoc_role_button
+      buttons += ansible_roles_button(args.first)
+
+      title_actions(button_group(buttons))
```
Comment (xprazak2) [inline app/helpers/foreman_ansible/hosts_helper_extensions.rb]: Maybe you could put the buttons into dropdown?

**Comment by xprazak2** [comment]
Nice work, I left a few comments inline. From looking at your `PlayHostRole` action, it has a lot in common with [PlayHostRoles](https://github.com/theforeman/foreman_ansible/blob/master/app/lib/actions/foreman_ansible/play_host_roles.rb), except for the roles and humanized name. Maybe you could make a change to pass the role_names into the `plan` method from the outside to make the class more generic:

```ruby
class PlayHostRoles < Actions::EntryAction
def plan(host, roles_names, proxy_selector = ::ForemanAnsible::ProxySelector.new,
               options = {})
        proxy = find_host_and_proxy(host, proxy_selector)
        inventory_creator = ::ForemanAnsible::InventoryCreator.new([host])
        playbook_creator = ::ForemanAnsible::PlaybookCreator.new(role_names)
        plan_delegated_action(proxy,
                              ::ForemanAnsibleCore::Actions::RunPlaybook,
                              :inventory => inventory_creator.to_hash.to_json,
                              :playbook => playbook_creator.roles_playbook,
                              :options => find_options.merge(options))
        plan_self
end
```
You could then `PlayHostRole < PlayHostRoles` and make your own implementation of humanized name or whatever you need. 

You would unfortunately need to refactor existing methods in controller a bit:

```ruby

          def play_roles
            role_names = @host.all_ansible_roles.map(&:name)
            @result = {
              :host => @host, :foreman_tasks => async_task(
                ::Actions::ForemanAnsible::PlayHostRoles, @host, role_names,
              )
            }

            render_message @result
          end
```

Of course, @dLobatog gets to have a final say in this ;-)

**Comment by Gerrit91** [comment]
Hey xprazak2, 

I saw that you did reviews recently and I appreciate your feedback and comments a lot. Unfortunately, I do not have the time to take care of the request momentarily. I also understand that, in the end, this feature maybe even does not really go into the direction that you want to go with the plugin. So, I am fine with closing this issue down without merging it in. No worries! :)

If you want to discuss this further, please feel free to contact me at any times.

Thanks again and regards,

--Gerrit

**Comment by dLobatog** [comment]
@Gerrit91 If you check the latest version, I think you ought to be able to do this through Remote execution 😄  - closing, but let me know if ther'es anything you think isn't covered in the latest version

**Pull request closed without merging**